### DISCUSSION TITLE
Middlware / other to record schema type of response

### DISCUSSION THREAD

**Discussion by joelcollinsdc** [thecodingmachine/graphqlite#701] [Q&A]
If I have a mutation that has a response schema of UnionShape1Shape2Shape3 (sort of like UnionSuccessTypeFailureTypeErrorsType), and I want to log out for any query / mutation, which concrete type the response was... would a custom middleware be the right thing for this?  How does one get access to the response schema in middleware?

**Comment by oojacoboo** [comment]
Back up a bit here and tell us your overall goals.  It sounds like you want to invalidate a session based on an error, and then return a custom response.  Is that correct?

**Reply by joelcollinsdc** [reply]
We want to record in our analytics platform which concrete type was returned from a mutation, so like if the response shape of our mutation is `UnionSuccessTypeFailureTypeErrorsType`, and it was a failure, we want to record/log that the response type was `FailureType` or `SuccessType`

**Reply by oojacoboo** [reply]
Why do you want to do that?  It'd be more helpful if you stepped back and stated your overall goals, instead of your solution you've come up with.  This is not the correct way to record errors.  That should be done with your exception handler.

**Reply by joelcollinsdc** [reply]
I'm stating the requirement that I have to implement based on my project.  My understanding is the exception handling process doesn't allow for custom schema defitnitions in the response.

**Reply by oojacoboo** [reply]
What do you mean by custom schema definitions in the response?  You may be stating what you actually need, or you may be stating the solution you've come up with based on your limited knowledge.

**Reply by joelcollinsdc** [reply]
Heres some psudeocode of what Im trying to do

```
#[Mutation]
	public function createFoo(
	CreateFooInput $createFooInput
): FooCreatedOutput | FailedToCreateFooOutput | ValidationErrors {
	if (!isValid()) {
		Logger::setParam('responseType', ValidationErrors::class);
		return new ValidationErrors(['example']);
	}


	try {
		$foo = new Foo();
		Logger::setParam('responseType', FooCreatedOutput::class);
		return new FooCreatedOutput('success!');
	} catch (\Throwable) {
		Logger::setParam('responseType', FailedToCreateFooOutput::class);
		return new FailedToCreateFooOutput();	
	}
}
```

What I was hoping to figure out is how to set `responseType` automatically

**Reply by oojacoboo** [reply]
Right, you're doing this all wrong.  You need to be using your global Exception handler for handling exceptions.  There is no reason to use a `try/catch` block here.  And your output type on this mutation, could just be FooCreatedOutput, or something more generic, but useful.  Often people will just return an `ID` of the newly created entity.

 https://www.php.net/manual/en/function.set-exception-handler.php

Use this function to set your callable for all exceptions and return a more uniform response.  Logging can be done there as well.

**Reply by aszenz** [reply]
I have a similar question kind of, how to document the error response as a union type, right now it has to be done separately for each query/mutation. 

What I see in graphql projects is that any query or mutation can return any kind of error so basically you cannot trust the output type on the client. 

A nice pattern is to categorize all the failures of a mutation in a union type , so the client is forced to handle these failure modes. This pattern is called `Errors as Values`.

One of the failure modes is generic error output for all mutations, this should be added to every mutations output type.

Basically how to override the graphql schema to add such a thing globally?

**Reply by aszenz** [reply]
Is it possible to add a decorator that decorates the final schema, it would be useful to extend and override the schema information generated by this lib. 

Also useful for adding consistent documentation across fields.

**Reply by oojacoboo** [reply]
Okay, I think I see what you're talking about here.

There is an extension object that's part of the response in the GraphQL spec.  Would this work for you?  There aren't any spec defined restrictions on this object.  You can basically format it however you need for your purposes.

In GraphQLite, you'd define your Webonyx `ServerConfig` and use `setErrorsHandler` and `setErrorFormatter`.  With your handler, you can customize the response, including the `extensions` key with whatever data you need.
